src/internal/fileserver/fileserver.go

```diff
@@ -0,0 +1,415 @@
+		// bytes -> human readable bytes
+		"humanizeBytes": func(x int64) string {
+			return humanize.Bytes(uint64(x))
+		},
+		// full path -> name of file
+		"basename": func(x string) string {
+			return path.Base(x)
+		},
+		// current URL path -> URL part after /pfs
+		"path": func(url string) string {
+			parts := strings.Split(url, "/")
+			if len(parts) > 2 {
+				return path.Join(parts[2:]...)
+			}
+			return url
+		},
+		// file info -> whether or not <url>/.. is valid
+		"hasParent": func(info *pfs.FileInfo) bool {
+			p := info.GetFile().GetPath()
+			return p != "/" && p != ""
+		},
+		// url -> url one level up; only valid if hasPrent == true
+		"parent": func(p string) string {
+			return path.Clean(path.Join(p, ".."))
+		},
+		// time -> rfc3339 string
+		"rfc3339": func(x time.Time) string {
+			return x.Format(time.RFC3339)
+		},
+	}).ParseFS(templateFS, "templates/*"))
+)
+
+// Server is an http.Handler that can download from and upload to PFS.
+type Server struct {
+	ClientFactory func(context.Context) *client.APIClient
+}
+
+type Request struct {
+	PachClient     *client.APIClient
+	Request        *http.Request
+	ResponseWriter http.ResponseWriter
+	RequestID      string
+	HTML           bool
+}
+
+// ServeHTTP implements http.Handler for the /pfs/ route.
+func (s *Server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	pachClient := httpauth.ClientWithToken(ctx, s.ClientFactory(ctx), req)
+	ctx = pachClient.Ctx()
+
+	ourRequest := &Request{
+		Request:        req,
+		ResponseWriter: w,
+		PachClient:     pachClient,
+		RequestID:      log.RequestID(ctx),
+		HTML:           acceptsHTML(req),
+	}
+
+	switch req.Method {
+	case http.MethodHead, http.MethodGet:
+		w.Header().Set("vary", constants.ContextTokenKey)
+		ourRequest.get(ctx)
+		return
+	default:
+		ourRequest.displayErrorf(ctx, http.StatusMethodNotAllowed, "unknown HTTP method %q", req.Method)
+		return
+	}
+}
+
+func (r *Request) get(ctx context.Context) {
+	req := r.Request
+	parts := strings.Split(req.URL.Path, "/")
+	switch {
+	case len(parts) == 5 && parts[4] != "":
+		// If parts[4] is "", then they are requesting what looks like a "list commits"
+		// page, that we don't implement.  Otherwise, they want the listing of the root
+		// directory, but forgot the trailing /.
+		http.Redirect(r.ResponseWriter, r.Request, r.Request.URL.Path+"/", http.StatusMovedPermanently)
+		return
+	case len(parts) >= 6: // /pfs/project/repo/commit|branch/path/to/some/file
+		parts[5] = path.Join(parts[5:]...)
+	default:
+		// Someday we can have a projects page, project repo page, repo commit page, etc.
+		// For now, only files and directories.
+		r.displayErrorf(ctx, http.StatusNotFound,
+			//                     0 1   2         3      4               5
+			"invalid URL; expecting /pfs/<project>/<repo>/<commit|branch>/<path...>, got %v",
+			strings.Join(parts, "/"))
+		return
+	}
+	if got, want := parts[1], "pfs"; got != want {
```

Comment:
very minor, but why not `if parts[1] != "pfs" {`?

Reply:
I think this style reduces errors where you check one thing and write something else in the error message.  If the URL becomes pFs or something, now you only have to change it in one place (instead of forgetting the error message and confusing people, if you just hard-coded it in the string and the == statement).